### PR TITLE
Ship and Platform clients

### PR DESCRIPTION
**Pull request by marccampbell**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -101,14 +101,14 @@ import (
 	"log"
 	"os"
 
-	"github.com/replicatedhq/replicated/client"
+	"github.com/replicatedhq/replicated/pkg/platform_client"
```
Comment (dexhorthy) [inline README.md]: I think underscores are discouraged in package names https://blog.golang.org/package-names

File: cli/cmd/root.go [hunk 2]
```diff
@@ -102,18 +106,38 @@ func Execute(stdin io.Reader, stdout io.Writer, stderr io.Writer) error {
...
 
-		app, err := api.GetApp(appSlugOrID)
+		appType, err := commonAPI.GetAppType(appSlugOrID)
 		if err != nil {
 			return err
 		}
-		runCmds.appID = app.Id
+		runCmds.appType = appType
+
+		if appType == "platform" {
```
Comment (dexhorthy) [inline cli/cmd/root.go]: yeah this feels like the kind of thing that would benefit from an interface but I don't really see a clear path to getting there yet

File: cli/cmd/root.go [hunk 3]
```diff
@@ -20,15 +23,16 @@ const (
...
+var shipOrigin = "https://g.replicated.com/graphql"
 
 func init() {
 	RootCmd.PersistentFlags().StringVar(&appSlugOrID, "app", "", "The app slug or app id to use in all calls")
 	RootCmd.PersistentFlags().StringVar(&apiToken, "token", "", "The API token to use to access your app in the Vendor API")
 
 	originFromEnv := os.Getenv("REPLICATED_API_ORIGIN")
 	if originFromEnv != "" {
-		apiOrigin = originFromEnv
+		platformOrigin = originFromEnv
```
Comment (emosbaugh) [inline cli/cmd/root.go]: do you wanna add something similar to here for ship? so we can test with staging?

File: client/app.go [hunk 4]
```diff
@@ -1,62 +1,59 @@
...
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return fmt.Errorf("DeleteApp (%s %s): %v", req.Method, endpoint, err)
-	}
-	resp.Body.Close()
-	if resp.StatusCode != http.StatusNoContent {
-		return fmt.Errorf("DeleteApp (%s %s): status %d", req.Method, endpoint, resp.StatusCode)
-	}
+func (c *Client) CreateApp(appOptions interface{}) (interface{}, error) {
+	return nil, nil
```
Comment (emosbaugh) [inline client/app.go]: i assume you will fill these in before merging?

File: pkg/ship_client/client.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	body, err := json.Marshal(requestObj)
+	if err != nil {
+		return errors.Wrap(err, "marshal body")
+	}
+
+	bodyReader := bytes.NewReader(body)
+
+	req, err := http.NewRequest("POST", c.GQLServer.String(), bodyReader)
+	if err != nil {
+		return errors.Wrap(err, "marshal body")
```
Comment (emosbaugh) [inline pkg/ship_client/client.go]: wrong message

File: pkg/ship_client/client.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", c.Token)
+
+	client := http.DefaultClient
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "marshal body")
```
Comment (emosbaugh) [inline pkg/ship_client/client.go]: wrong message

File: pkg/ship_client/client.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		return errors.New("nil response from gql")
+	}
+
+	if resp.Body == nil {
+		return errors.New("nil response.Body from gql")
+	}
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrap(err, "marshal body")
```
Comment (emosbaugh) [inline pkg/ship_client/client.go]: wrong message

File: pkg/ship_client/app.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func (c *GraphQLClient) GetApp(appID string) (*types.App, error) {
+	apps, err := c.ListApps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, app := range apps {
+		if app.App.ID == appID || app.App.Slug == appID {
```
Comment (emosbaugh) [inline pkg/ship_client/app.go]: seems like there should be a better way

File: client/client.go [hunk 9]
```diff
@@ -1,115 +1,44 @@
...
-	return c
+	return client
 }
 
-func NewHTTPClient(origin string, apiKey string) Client {
-	c := &HTTPClient{
-		apiKey:    apiKey,
-		apiOrigin: origin,
+func (c *Client) GetAppType(appID string) (string, error) {
+	platformApps, err := c.PlatformClient.ListApps()
```
Comment (emosbaugh) [inline client/client.go]: GetApp() ?

File: cli/cmd/root.go (hunk 2 above)
Comment (marccampbell) [reply cli/cmd/root.go]: agreed.. i'm going to add some more implementation to ship methods, and look for opportunity for a pattern to emerge

File: client/app.go (hunk 4 above)
Comment (marccampbell) [reply client/app.go]: Yes, but not a blocker here. The CreateApp function in the CLI still calls the platformapi method directly. My intent isn't to add all ship functionality in 1 PR, but to start adding it, as needed.

**Pull request merged**